README
This software calculates a password for resetting the clock in the
Game Boy Color games Pokémon Gold and Silver. It can work with Pokémon Crystal
aswell, but it requires a GameShark.

Where to Get the Information
--------------------------------------------------------------------------------
To get the player-character name, trainer ID number, and amount of held money,
enter the main menu by pressing START. Then select the option labelled with the
player character name (this should be above SAVE). The player character name,
trainer ID number, and amount of held money are at the top of the screen.

Using the Password
--------------------------------------------------------------------------------
In Gold and Silver, at the title screen, press DOWN+SELECT+B. Select YES to
reset the clock. Enter the password at the password prompt. It resets the clock
upon receiving the correct password. Next time you select CONTINUE at the main
menu, it will ask for the day of week and the time.

In Crystal, it is not possible to reset the clock legitimately. If you are
willing and able to use a GameShark, the code is 010464CF. After entering this
GameShark code, press A at the title screen. It will then ask if you want to
reset the clock. At this point, remove the GameShark code and then proceed
normally.

I used this page as a reference in the making of this software:
http://web.archive.org/web/20090902012638/http://geocities.com/thelegendarydogs/faqs/gsc/time/timereset.htm


CHANGES
Version 0.1, 2011-09-30
 - First official release
Version 0.2, 2011-09-30
 - More efficient calculation of name character values
 - Notice in player-character name prompt that name is case-sensitive and to use
   { and } for PK and MN, respectively
Version 0.3, 2011-10-01
 - Fixed space bug (it would throw an error if the player-character name
   contained a space, which is a valid character in the player-character name
   in the actual game)
 - Improved error handling
 - Padding of ID number and password if less than 5 digits long
Version '0.3w', 2020-05-10
 - Added a Python 2.7 backport, which I actually wrote in nano on an iPhone 4S
   in 2015. Python 3 version remains unchanged.
 - Improved documentation (README.md).

pkmn-gs-clockpassword.py
#!/usr/bin/env python

# Pokémon Gold/Silver Clock Reset Password Generator
# Version 0.3
# Copyright 2011 woddfellow2 | http://wlair.us.to/
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# 
# TODO:
# - Command-line options

import sys;

# Collect player character name, trainer ID number, and amount of held money,
# also error handling:
name = input("Enter name of player character (case-sensitive; use { for PK"
             " and } for MN):\n> ");
if len(name) > 7:
    print("ER001: Name must be only 7 characters long.", file=sys.stderr);
    exit();
if name == "":
    print("ER005: Name must not be blank.", file=sys.stderr);
    exit();
try:
    trainerid = int(input("Enter ID number:\n> "));
except ValueError:
    print("ER006: Trainer ID given was blank or not a number.",
          file=sys.stderr);
    exit();
if 0 <= trainerid >= 65536:
    print("ER002: Trainer ID must be a 5-digit number from 00000 to 65536.",
          file=sys.stderr);
    exit();
try:
    money = int(input("Enter amount of held money:\n> "));
except ValueError:
    print("ER007: Amount of held money given was blank or not a number.",
          file=sys.stderr);
    exit();
if 0 <= money >= 999999:
    print("ER003: Amount of money must be a 0- to 6-digit number.",
          file=sys.stderr);
    exit();

# Name
# Translate first 5 characters of name into values:
name_chars = { "A": 128, "B": 129, "C": 130, "D": 131, "E": 132, "F": 133,
               "G": 134, "H": 135, "I": 136, "J": 137, "K": 138, "L": 139,
               "M": 140, "N": 141, "O": 142, "P": 143, "Q": 144, "R": 145,
               "S": 146, "T": 147, "U": 148, "V": 149, "W": 150, "X": 151,
               "Y": 152, "Z": 153, "(": 154, ")": 155, ":": 156, ";": 157,
               "[": 158, "]": 159, "a": 160, "b": 161, "c": 162, "d": 163,
               "e": 164, "f": 165, "g": 166, "h": 167, "i": 168, "j": 169,
               "k": 170, "l": 171, "m": 172, "n": 173, "o": 174, "p": 175,
               "q": 176, "r": 177, "s": 178, "t": 179, "u": 180, "v": 181,
               "w": 182, "x": 183, "y": 184, "z": 185, "{": 225, "}": 226,
               "-": 227, "?": 230, "!": 231, ".": 232, "*": 241, "/": 243,
               ",": 244, " ": 0 }

for k, v in name_chars.items():
    if name[0] == k:
        name_char1 = v;
    if len(name) >= 2:
        if name[1] == k:
            name_char2 = v;
    if len(name) >= 3:
        if name[2] == k:
            name_char3 = v;
    if len(name) >= 4:
        if name[3] == k:
            name_char4 = v;
    if len(name) >= 5:
        if name[4] == k:
            name_char5 = v;

try:
    if len(name) == 1:
        name_total = int(name_char1);
    if len(name) == 2:
        name_total = int(name_char1) + int(name_char2);
    if len(name) == 3:
        name_total = int(name_char1) + int(name_char2) + int(name_char3);
    if len(name) == 4:
        name_total = int(name_char1) + int(name_char2) + int(name_char3) \
        + int(name_char4);
    if len(name) >= 5:
        name_total = int(name_char1) + int(name_char2) + int(name_char3) \
        + int(name_char4) + int(name_char5);
    
    if len(name) < 5:
        name_total = int(name_total) + 80;
except NameError:
    print("ER004: Name given contains one or more invalid characters.",
          file=sys.stderr);
    exit();

# Money
money_byte1 = int(money) / 65536;
money_byte2 = int(money) / 256 % 256;
money_byte3 = int(money) % 256;

money_total = int(money_byte1) + int(money_byte2) + int(money_byte3);

# Trainer ID
trainerid_byte1 = int(trainerid) / 256;
trainerid_byte2 = int(trainerid) % 256;

trainerid_total = int(trainerid_byte1) + int(trainerid_byte2);

# Password
password = int(name_total) + int(money_total) + int(trainerid_total);

# Output result
print(); # Newline to make it look better
print("%s/%s, %d units of currency" % (name, str(trainerid).zfill(5), money));
print("Password:", str(int(password)).zfill(5));


pkmn-gs-clockpassword_2.7.py
#!/usr/bin/env python2.7
# coding=utf-8
# Pokémon Gold/Silver Clock Reset Password Generator
# Version 0.3
# Copyright 2011 woddfellow2 | http://wlair.us.to/
# Python 2.7 "port" also copyright 2015 Wyatt Ward.
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# 
# TODO:
# - Command-line options

import sys;

# Collect player character name, trainer ID number, and amount of held money,
# also error handling:
name = raw_input("Enter name of player character (case-sensitive; use { for PK and } for MN):\n> ");
if len(name) > 7:
    sys.stderr.write("ER001: Name must be only 7 characters long.\n");
    exit();
if name == "":
    sys.stderr.write("ER005: Name must not be blank.\n");
    exit();
try:
    trainerid = int(raw_input("Enter ID number:\n> "));
except ValueError:
    sys.stderr.write("ER006: Trainer ID given was blank or not a number.\n");
    exit();
if 0 <= trainerid >= 65536:
    sys.stderr.write("ER002: Trainer ID must be a 5-digit number from 00000 to 65536.\n");
    exit();
try:
    money = int(raw_input("Enter amount of held money:\n> "));
except ValueError:
    sys.stderr.write("ER007: Amount of held money given was blank or not a number.\n");
    exit();
if 0 <= money >= 999999:
    sys.stderr.write("ER003: Amount of money must be a 0- to 6-digit number.\n");
    exit();

# Name
# Translate first 5 characters of name into values:
name_chars = { "A": 128, "B": 129, "C": 130, "D": 131, "E": 132, "F": 133,
               "G": 134, "H": 135, "I": 136, "J": 137, "K": 138, "L": 139,
               "M": 140, "N": 141, "O": 142, "P": 143, "Q": 144, "R": 145,
               "S": 146, "T": 147, "U": 148, "V": 149, "W": 150, "X": 151,
               "Y": 152, "Z": 153, "(": 154, ")": 155, ":": 156, ";": 157,
               "[": 158, "]": 159, "a": 160, "b": 161, "c": 162, "d": 163,
               "e": 164, "f": 165, "g": 166, "h": 167, "i": 168, "j": 169,
               "k": 170, "l": 171, "m": 172, "n": 173, "o": 174, "p": 175,
               "q": 176, "r": 177, "s": 178, "t": 179, "u": 180, "v": 181,
               "w": 182, "x": 183, "y": 184, "z": 185, "{": 225, "}": 226,
               "-": 227, "?": 230, "!": 231, ".": 232, "*": 241, "/": 243,
               ",": 244, " ": 0 }

for k, v in name_chars.items():
    if name[0] == k:
        name_char1 = v;
    if len(name) >= 2:
        if name[1] == k:
            name_char2 = v;
    if len(name) >= 3:
        if name[2] == k:
            name_char3 = v;
    if len(name) >= 4:
        if name[3] == k:
            name_char4 = v;
    if len(name) >= 5:
        if name[4] == k:
            name_char5 = v;

try:
    if len(name) == 1:
        name_total = int(name_char1);
    if len(name) == 2:
        name_total = int(name_char1) + int(name_char2);
    if len(name) == 3:
        name_total = int(name_char1) + int(name_char2) + int(name_char3);
    if len(name) == 4:
        name_total = int(name_char1) + int(name_char2) + int(name_char3) \
        + int(name_char4);
    if len(name) >= 5:
        name_total = int(name_char1) + int(name_char2) + int(name_char3) \
        + int(name_char4) + int(name_char5);
    
    if len(name) < 5:
        name_total = int(name_total) + 80;
except NameError:
    sys.stderr.write("ER004: Name given contains one or more invalid characters.\n");
    exit();

# Money
money_byte1 = int(money) / 65536;
money_byte2 = int(money) / 256 % 256;
money_byte3 = int(money) % 256;

money_total = int(money_byte1) + int(money_byte2) + int(money_byte3);

# Trainer ID
trainerid_byte1 = int(trainerid) / 256;
trainerid_byte2 = int(trainerid) % 256;

trainerid_total = int(trainerid_byte1) + int(trainerid_byte2);

# Password
password = int(name_total) + int(money_total) + int(trainerid_total);

# Output result
#print(); # Newline to make it look better (removed in python 3 to 2 down-conversion
print(""); # Newline to make it look better (removed in python 3 to 2 down-conversion
print("%s/%s, %d units of currency" % (name, str(trainerid).zfill(5), money));
#print("Password:", str(int(password)).zfill(5));
print("Password: %s" % (str(int(password)).zfill(5)));
#print("Password: %s" % str(int(password)));
